torchrl/data/tensor_specs.py

```diff
@@ -997,6 +992,72 @@ def _project(self, val: torch.Tensor) -> torch.Tensor:
         return torch.cat([super()._project(_val) for _val in vals], -1)
 
 
+class DiscreteTensorSpec(TensorSpec):
+    """A discrete tensor spec.
+
+    An alternative to OneHotTensorSpec for categorical variables in TorchRL. Instead of
+    using multiplication, categorical variables perform indexing which can speed up
+    computation and reduce memory cost for large categorical variables.
+        >>> batch, size = 3, 4
+        >>> action_value = torch.arange(batch*size)
+        >>> action_value = action_value.view(batch, size).to(torch.float)
+        >>> action = torch.argmax(action_value, dim=-1).to(torch.long)
+        >>> chosen_action_value = action_value[range(batch), action]
+        >>> print(chosen_action_value)
+        tensor([ 3.,  7., 11.])
+
+    Args:
+        n (int): number of possible outcomes.
+        shape: (torch.Size, optional): shape of the variable, default is "(1,)".
+        device (str, int or torch.device, optional): device of the tensors.
+        dtype (str or torch.dtype, optional): dtype of the tensors.
+
+    """
+
+    shape: torch.Size
+    space: DiscreteBox
+    device: torch.device = torch.device("cpu")
+    dtype: torch.dtype = torch.float
+    domain: str = ""
+
+    def __init__(
+        self,
+        n: int,
+        shape: Optional[torch.Size] = torch.Size((1,)),
```

Comment:
IIRC we once had a linting issue with this (that seems to have disappeared now).
This is why we usually have a `None` and if `None` then shape is replaced by `Size([1])`.

Reply:
Noted, changed to that